fossology/fossology
# feat(buildsystem): add CMake Buildsytem

## Description

This pull request adds a new build system based on CMake to this project. This is a big upgrade over decade-old Makefiles. With the new build system, we will be able to build/install/package FOSSology faster and support for many platforms and another build system. 

### Changes

- Added CMake configuration for building, installing, testing, and packaging all the agents
- Refactored source directories to better integrate with the new build system
- Reworked the shell scripts and generated source files to make them more compatible with CMake as well as better in terms of overall compatibility with the latest tools.
- Refactored test source 

## How to test

- Check out the pull request branch 

 ```bash
    git fetch https://github.com/avinal/fossology avinal/feat/buildsystem:buildsystem
    git checkout buildsystem
```

- Make sure you have CMake and Ninja installed

```bash
  sudo apt install cmake ninja-build
```

- Configure CMake, you may add additional variables to modify build
  - `-DOFFLINE=ON` - stops vendor generation
  - `-DAGENTS="agent1;agent2;...;<agent>"` - only configure these agents
  - `-DTESTING=ON` - enables configuring for tests
  - `-DCMAKE_INSTALL_PREFIX=<path>` - changes install prefix for the project
  - `-DMONOPACK=ON` - ununpack and adj2nest are packaged seperately
  - `-GNinja` - uses Ninja Generator instead of  Unix Makefiles
  - `-DCMAKE_BUILD_TYPE=<build-type>` - choose from `Debug`, `Release`, `RelWithDebInfo`, `MinSizeRel`
 
```bash
  mkdir build && cd build
  cmake .. <add-variables-here-if-any>
```

- Run Make or Ninja

```bash
  # For Unix Makefiles, by default, runs without parallel build jobs
  make -j<no-of-parallel-jobs>
  # For Ninja Generator, by default, runs parallel build jobs
  ninja -j<no-of-parallel-jobs>
  # common for all
  cmake --build .
```

- Installing, may need to add `sudo` before the commands

```bash
  # For Unix Makefiles
  make install
  # For Ninja Generator
  ninja install
```

- Testing, on local please consider switching to `fossy` user before running tests

```bash
  # For Unix Makefiles
  make test
  # For Ninja Generator
  ninja test
  # dedicated ctest command *see `ctest --help` for more options*
  ctest --parallel n
```

- Packaging

```bash
  # For Unix Makefiles
  make package
  # For Ninja Generator
  ninja package
```

- Follow the rest of the installation from [here](https://github.com/fossology/fossology/wiki/Install-from-Source#4-run-the-postinstall-script)
- To list all the possible targets 

```bash
  # For Unix Makefiles
  make help
  # For Ninja Generator
  ninja -t targets
```

## Known Bugs/Issue

- While packaging with `make` symbolic links in the packages are accompanied with dereferenced targets too, prefer `ninja` for packaging
- Coverage build may fail for some agents

### Issues to create before merge
- [ ] Fix tests for scheduler, ununpack and delagent units tests.
- [ ] Add requirements for CMake version > 3.13 in Wiki
- [ ] Update install from source instructions in Wiki
- [ ] Add packaging instructions to use `-DCMAKE_INSTALL_PREFIX:PATH="/usr" -DFO_SYSCONFDIR:PATH="/etc/fossology"` before packaging to fix paths.


For More Information please see the blog: https://gsoc.avinal.space and [wiki](https://github.com/avinal/fossology/wiki)

Signed-off-by: Avinal Kumar <avinal.xlvii@gmail.com>


<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2075"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
